lasy/utils/mode_decomposition.py

```diff
@@ -0,0 +1,143 @@
+    -------
+    weights: dict of floats
+        A dictionary of floats corresponding to the weights of each mode
+        in the decomposition. The keys of the dictionary are tuples
+        corresponding to (`n_x`,`n_y`)
+
+    """
+
+    # Check if the provided laserProfile is a full laser profile or a
+    # transverse profile.
+
+    assert isinstance(
+        laserProfile, TransverseProfile
+    ), "laserProfile must be an instance of TransverseProfile"
+
+    # Here we need to define a grid size to use for the calculation
+    # By default we will use N x N grid where N_pts = 500
+    N_pts = 500
+
+    # Get the field, sensible spatial bounds for the profile
+    lo = [None, None]
+    hi = [None, None]
+    if isinstance(laserProfile, TransverseProfileFromData):
+        lo[0] = laserProfile.field_interp.grid[0].min() + laserProfile.x_offset
+        lo[1] = laserProfile.field_interp.grid[1].min() + laserProfile.x_offset
+        hi[0] = laserProfile.field_interp.grid[0].max() + laserProfile.y_offset
+        hi[1] = laserProfile.field_interp.grid[1].max() + laserProfile.y_offset
+
+    else:
+        lo[0] = -laserProfile.w0 * 5 + laserProfile.x_offset
+        lo[1] = -laserProfile.w0 * 5 + laserProfile.x_offset
+        hi[0] = laserProfile.w0 * 5 + laserProfile.x_offset
+        hi[1] = laserProfile.w0 * 5 + laserProfile.x_offset
+
+    # Define spatial arrays
+    x = np.linspace(lo[0], hi[0], N_pts)
+    y = np.linspace(lo[1], hi[1], N_pts)
+    X, Y = np.meshgrid(x, y)
+    dx = x[2] - x[1]
+    dy = y[2] - y[1]
+
+    # Get the field on this grid
+    field = laserProfile.evaluate(X, Y)
+
+    # Get estimate of w0
+    w0 = estimate_best_HG_waist(x, y, field)
+
+    # Next we loop over the modes and calculate the relevant weights
+    weights = {}
+    for i in range(n_x_max):
+        for j in range(n_y_max):
+            HGMode = HermiteGaussianTransverseProfile(w0, i, j)
+            coef = np.sum(field * HGMode.evaluate(X, Y)) * dx * dy  # modalDecomposition
+            if math.isnan(coef):
+                coef = 0
+            weights[(i, j)] = coef
+
+    return weights, w0
+
+
+def estimate_best_HG_waist(x, y, field):
+    """
+    Estimate the waist that maximises the weighting of the first mode.
+
+    Calculates a D4Sigma waist as a first estimate and then tests multiple
+    gaussians with waists around this value to determine which has the best
+    overlap with the provided intensity profile. The aim here is to maximise
+    the energy in the fundamental mode of the reconstruction and so to avoid
+    a decomposition with significant higher-order modal content.
+
+    Parameters
+    ---------
+    x,y : 1D numpy arrays
+        representing the x and y axes on which the intensity profile is defined.
+
+    field : 2D numpy array representing the field (not the laser intensity).
+        the laser field profile in a 2D slice.
+
+    """
+
+    dx = x[2] - x[1]
+    dy = y[2] - y[1]
+
+    assert dx == dy
+
```

Comment:
Maybe this is required? I didn't test this assert
```suggestion
    assert field.shape == (x.size, y.size)
```

Reply:
I don't think it's necessary for x and y to be the same length, in fact for most experimental data this won't be the case, cameras tend to have rectangular chips. But I do think we start to have problems if dx and dy are not the same

Reply:
That's not testing `x.size == y.size`.

Reply:
yes, i misunderstood the original point. I think this makes sense